src/main/java/com/machine/learning/classifier/#ID3.java#

```diff
@@ -0,0 +1,303 @@
+	    if (dataPoint.get(curDT.attributeIndex).equals(curDT.attributeValue)) {
+		curDT = curDT.pos;
+	    } else {
+		curDT = curDT.neg;
+	    }
+	}
+	
+	return curDT.clazz;
+    }
+
+    public DecisionTree constructDT(List<DataPoint> remainingData) {
+	ArrayList<String> classes = new ArrayList<>();
+        remainingData.get(0).getClassLabel().get();
+	//ArrayList<double> classProportion = new ArrayList<>();
+	boolean allOneClazz = true;
+	ArrayList<ArrayList<String>> usedAttrValues = new ArrayList<>();
+
+	for(int i = 0; i < remainingData.get(0).getData().get().size(); i++) {
+	    usedAttrValues.set(i, new ArrayList<>());
+	}
+
+	for (DataPoint dataPoint : remainingData) {
+	    if (allOneClazz && !dataPoint.getClassLabel().get().equals(classes.get(0))) {
+		allOneClazz = false;
+	    }
+	    if (!classes.contains(dataPoint.getClassLabel().get())) {
+		classes.add(dataPoint.getClassLabel().get());
+		//classProportion.add(1.0);
+	    } else {
+		int index = classes.indexOf(dataPoint.getClassLabel().get());
+		//classProportion.set(index, classProportion.get(index)++);
+	    }
+	    /*
+	    ArrayList<String> curData = dataPoint.getData().get();
+	    for(int j = 0; j < curData.size(); j++){
+		if (!usedAttrValues.get(j).contains(curData.get(j))) {
+		    usedAttrValues.get(j).add(curData.get(j));
+		}
+	    }
+	    */
+	}
+
+	if (allOneClazz) {
+	    return new DecisionTree(classes.get(0));
+	}
+
+	// for (int i = 0;  i < classProportion.size(); i++) {
+	//     classProportion.set(i, classProportion.get(i)/remainingData.size());
+	// }
+	
+	// for (DataPoint dataPoint : remainingData) {
+	//     for (String attr : dataPoint.getData().get()) {
+	// 	//store somewhere//
+	// 	calculateEntropy(attr, remainingData);
+	//     }
+	// }
+	int attrIndex = 0;
+	String attributeValue;
+	double minEntropy = Double.MAX_VALUE;
+	ArrayList<DataPoint> posData = new ArrayList<>();
+	ArrayList<DataPoint> negData = new ArrayList<>();
+
+	for (int i = 0; i < usedAttrValues.size(); i++) {
+	    for (String attrValue : usedAttrValues.get(i)) {
+		for (DataPoint dataPoint : remainingData) {
+		    posData.clear();
+		    negData.clear();
+
+		    if (dataPoint.getData().get().contains(attrValue)) {
+			posData.add(dataPoint);
+		    } else {
+			negData.add(dataPoint);
+		    }
+		    double entropy = Math.min(calculateEntropy(posData), calculateEntropy(negData));
+		    if(entropy < minEntropy){
+			minEntropy = entropy;
+			attrIndex = i;
+			attributeValue = attrValue;
+		    }
+		}
+	    }
+	}
+
+	posData.clear();
+	negData.clear();
+
+	//get attribute (attrValue) and index (attrIndex) of lowest entropy value
+	for (DataPoint dataPoint : remainingData) {
+	    if (dataPoint.getData().get().contains(attributeValue)) {
+		posData.add(dataPoint);
+	    } else {
+		negData.add(dataPoint);
+	    }
+	}
+	
+	DecisionTree cur = new DecisionTree(attrIndex, attributeValue);
+	cur.pos = constructDT(posData);
+	cur.neg = constructDT(negData);
+
+	return cur;
+	
+    }
```

Comment:
I won't make you do this, but
this method is super long so generally you would break out blocks where possible. It makes code easier to follow and more self documenting